server/clustering_test.go

```diff
@@ -644,6 +644,159 @@ func TestClusteringLogSnapshotRestore(t *testing.T) {
 	}
 }
 
+func TestClusteringLogSnapshotRestoreSubAcksPending(t *testing.T) {
+	cleanupDatastore(t)
+	defer cleanupDatastore(t)
+	cleanupRaftLog(t)
+	defer cleanupRaftLog(t)
+
+	// For this test, use a central NATS server.
+	ns := natsdTest.RunDefaultServer()
+	defer ns.Shutdown()
+
+	// Configure first server
+	s1sOpts := getTestDefaultOptsForClustering("a", []string{"b", "c"})
+	s1sOpts.Clustering.TrailingLogs = 0
+	s1 := runServerWithOpts(t, s1sOpts, nil)
+	defer s1.Shutdown()
+
+	// Configure second server.
+	s2sOpts := getTestDefaultOptsForClustering("b", []string{"a", "c"})
+	s2sOpts.Clustering.TrailingLogs = 0
+	s2 := runServerWithOpts(t, s2sOpts, nil)
+	defer s2.Shutdown()
+
+	// Configure third server.
+	s3sOpts := getTestDefaultOptsForClustering("c", []string{"a", "b"})
+	s3sOpts.Clustering.TrailingLogs = 0
+	s3 := runServerWithOpts(t, s3sOpts, nil)
+	defer s3.Shutdown()
+
+	servers := []*StanServer{s1, s2, s3}
+	for _, s := range servers {
+		checkState(t, s, Clustered)
+	}
+
+	// Wait for metadata leader to be elected.
+	getMetadataLeader(t, 10*time.Second, servers...)
+
+	// Create a client connection.
+	sc, err := stan.Connect(clusterName, clientName)
+	if err != nil {
+		t.Fatalf("Expected to connect correctly, got err %v", err)
+	}
+	defer sc.Close()
+
+	// Publish a message (this will create the channel and form the Raft group).
+	channel := "foo"
+	publishWithRetry(t, sc, channel, []byte("1"))
+
+	// Wait for leader to be elected.
+	leader := getChannelLeader(t, channel, 10*time.Second, servers...)
+
+	// Create a subscription.
+	ch := make(chan *stan.Msg, 1)
+	_, err = sc.Subscribe(channel, func(msg *stan.Msg) {
+		// Do not ack.
+		ch <- msg
+	}, stan.DeliverAllAvailable(), stan.SetManualAckMode(), stan.AckWait(time.Second), stan.MaxInflight(1))
+	if err != nil {
+		t.Fatalf("Unexpected error on subscribe: %v", err)
+	}
+
+	// Verify we received the message.
+	select {
+	case msg := <-ch:
+		assertMsg(t, msg.MsgProto, []byte("1"), 1)
+	case <-time.After(2 * time.Second):
+		t.Fatal("expected msg")
+	}
+
+	// Kill a follower.
+	var follower *StanServer
+	for _, s := range servers {
+		if leader != s {
+			follower = s
+			break
+		}
+	}
+	follower.Shutdown()
```

Comment:
I was thinking that the follower should be done before the events (pub/sub/acks) occur, otherwise you are not checking that the follower really recover from snapshot, it could have had the state from before being shutdown, no?

Reply:
Fixed.